Better diagnostics for broken libclang

Commit bfe196, latest install_prereqs.sh has been run, this is on Ubuntu 20:

```
└─ $ ▶ bazel build -c dbg manipulation/util:geometry_inspector
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
INFO: Analyzed target //manipulation/util:geometry_inspector (173 packages loaded, 8271 targets configured).
INFO: Found 1 target...
ERROR: /home/drum/drake-new/bindings/pydrake/BUILD.bazel:57:37: Action bindings/pydrake/documentation_pybind.h failed: (Exit 1): mkdoc failed: error executing command bazel-out/host/bin/tools/workspace/pybind11/mkdoc -DDRAKE_COMMON_SYMBOLIC_DETAIL_HEADER -DEIGEN_MPL2_ONLY -DHAVE_CSTDDEF '-DFMT_HEADER_ONLY=1' '-DFMT_NO_FMT_STRING_ALIAS=1' -DSPDLOG_COMPILED_LIB ... (remaining 743 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
external/eigen/include/_usr_include_eigen3/Eigen/Core:82:12: fatal error: 'new' file not found
Traceback (most recent call last):
  File "/home/drum/.cache/bazel/_bazel_drum/2059ccae995efcc54d793ceb88203d5b/sandbox/linux-sandbox/1197/execroot/drake/bazel-out/host/bin/tools/workspace/pybind11/mkdoc.runfiles/drake/tools/workspace/pybind11/mkdoc.py", line 841, in <module>
    main()
  File "/home/drum/.cache/bazel/_bazel_drum/2059ccae995efcc54d793ceb88203d5b/sandbox/linux-sandbox/1197/execroot/drake/bazel-out/host/bin/tools/workspace/pybind11/mkdoc.runfiles/drake/tools/workspace/pybind11/mkdoc.py", line 801, in main
    raise RuntimeError(
RuntimeError: Parsing headers using the clang library failed with 0 error(s) and 1 fatal error(s)
----------------
Note: The failure of target //tools/workspace/pybind11:mkdoc (with exit code 1) may have been caused by the fact that it is running under Python 3 instead of Python 2. Examine the error to determine if that appears to be the problem. Since this target is built in the host configuration, the only way to change its version is to set --host_force_python=PY2, which affects the entire build.

If this error started occurring in Bazel 0.27 and later, it may be because the Python toolchain now enforces that targets analyzed as PY2 and PY3 run under a Python 2 and Python 3 interpreter, respectively. See https://github.com/bazelbuild/bazel/issues/7899 for more information.
```

Looks like the error has something to do with Eigen, not python.

FYI As of #15798, here's the error message that comes out on Focal if gcc-10 is installed without g++-10 also being installed.  Maybe it will help with the googles.

```
ERROR: .../bindings/pydrake/BUILD.bazel:57:37: Action bindings/pydrake/documentation_pybind.h failed: (Exit 1): mkdoc failed: error executing command bazel-out/host/bin/tools/workspace/pybind11/mkdoc -DDRAKE_COMMON_SYMBOLIC_DETAIL_HEADER -DEIGEN_MPL2_ONLY -DHAVE_CSTDDEF '-DFMT_HEADER_ONLY=1' '-DFMT_NO_FMT_STRING_ALIAS=1' -DSPDLOG_COMPILED_LIB ... (remaining 742 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-opt/bin/bindings/pydrake/documentation_pybind.h.tmp_artifacts/mkdoc_glue.h:1:10: fatal error: 'optional' file not found
clang version 9.0.1-12 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: 
 "clang" "-cc1" "-triple" "x86_64-pc-linux-gnu" "-fsyntax-only" "-disable-free" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "foo" "-mrelocation-model" "static" "-mthread-model" "posix" "-mdisable-fp-elim" "-fallow-editor-placeholders" "-fmath-errno" "-masm-verbose" "-mconstructor-aliases" "-munwind-tables" "-fuse-init-array" "-target-cpu" "x86-64" "-dwarf-column-info" "-debugger-tuning=gdb" "-resource-dir" "lib/clang/9.0.1" "-isystem" "lcmtypes" "-isystem" "bazel-out/k8-opt/bin/lcmtypes" "-isystem" "external/lcm" "-isystem" "bazel-out/k8-opt/bin/external/lcm" "-isystem" "external/lcm/lcm" "-isystem" "bazel-out/k8-opt/bin/external/lcm/lcm" "-isystem" "external/eigen/include/_usr_include_eigen3" "-isystem" "bazel-out/k8-opt/bin/external/eigen/include/_usr_include_eigen3" "-isystem" "external/fmt/include" "-isystem" "bazel-out/k8-opt/bin/external/fmt/include" "-isystem" "external/optitrack_driver/lcmtypes" "-isystem" "bazel-out/k8-opt/bin/external/optitrack_driver/lcmtypes" "-isystem" "external/clp/include/_usr_include_coin" "-isystem" "bazel-out/k8-opt/bin/external/clp/include/_usr_include_coin" "-isystem" "external/nlopt/include" "-isystem" "bazel-out/k8-opt/bin/external/nlopt/include" "-isystem" "external/ibex/opt/libibex/2.7.4/include" "-isystem" "bazel-out/k8-opt/bin/external/ibex/opt/libibex/2.7.4/include" "-isystem" "external/ibex/opt/libibex/2.7.4/include/ibex" "-isystem" "bazel-out/k8-opt/bin/external/ibex/opt/libibex/2.7.4/include/ibex" "-isystem" "external/ibex/opt/libibex/2.7.4/include/ibex/3rd" "-isystem" "bazel-out/k8-opt/bin/external/ibex/opt/libibex/2.7.4/include/ibex/3rd" "-isystem" "external/ipopt/include/_usr_include_coin" "-isystem" "bazel-out/k8-opt/bin/external/ipopt/include/_usr_include_coin" "-isystem" "external/vtk/include/vtk-8.2" "-isystem" "bazel-out/k8-opt/bin/external/vtk/include/vtk-8.2" "-isystem" "external/double_conversion/include" "-isystem" "bazel-out/k8-opt/bin/external/double_conversion/include" "-isystem" "external/libjpeg/include" "-isystem" "bazel-out/k8-opt/bin/external/libjpeg/include" "-isystem" "external/libpng/include/_usr_include_libpng16" "-isystem" "bazel-out/k8-opt/bin/external/libpng/include/_usr_include_libpng16" "-isystem" "external/libtiff/include/_usr_include_x86_64-linux-gnu" "-isystem" "bazel-out/k8-opt/bin/external/libtiff/include/_usr_include_x86_64-linux-gnu" "-isystem" "external/opengl/include" "-isystem" "bazel-out/k8-opt/bin/external/opengl/include" "-isystem" "external/yaml_cpp/include" "-isystem" "bazel-out/k8-opt/bin/external/yaml_cpp/include" "-iquote" "." "-iquote" "bazel-out/k8-opt/bin" "-iquote" "external/lcm" "-iquote" "bazel-out/k8-opt/bin/external/lcm" "-iquote" "external/eigen" "-iquote" "bazel-out/k8-opt/bin/external/eigen" "-iquote" "external/fmt" "-iquote" "bazel-out/k8-opt/bin/external/fmt" "-iquote" "external/optitrack_driver" "-iquote" "bazel-out/k8-opt/bin/external/optitrack_driver" "-iquote" "external/clp" "-iquote" "bazel-out/k8-opt/bin/external/clp" "-iquote" "external/nlopt" "-iquote" "bazel-out/k8-opt/bin/external/nlopt" "-iquote" "external/ibex" "-iquote" "bazel-out/k8-opt/bin/external/ibex" "-iquote" "external/libblas" "-iquote" "bazel-out/k8-opt/bin/external/libblas" "-iquote" "external/liblapack" "-iquote" "bazel-out/k8-opt/bin/external/liblapack" "-iquote" "external/zlib" "-iquote" "bazel-out/k8-opt/bin/external/zlib" "-iquote" "external/ipopt" "-iquote" "bazel-out/k8-opt/bin/external/ipopt" "-iquote" "external/vtk" "-iquote" "bazel-out/k8-opt/bin/external/vtk" "-iquote" "external/double_conversion" "-iquote" "bazel-out/k8-opt/bin/external/double_conversion" "-iquote" "external/liblz4" "-iquote" "bazel-out/k8-opt/bin/external/liblz4" "-iquote" "external/liblzma" "-iquote" "bazel-out/k8-opt/bin/external/liblzma" "-iquote" "external/libjpeg" "-iquote" "bazel-out/k8-opt/bin/external/libjpeg" "-iquote" "external/libpng" "-iquote" "bazel-out/k8-opt/bin/external/libpng" "-iquote" "external/libtiff" "-iquote" "bazel-out/k8-opt/bin/external/libtiff" "-iquote" "external/glew" "-iquote" "bazel-out/k8-opt/bin/external/glew" "-iquote" "external/opengl" "-iquote" "bazel-out/k8-opt/bin/external/opengl" "-iquote" "external/x11" "-iquote" "bazel-out/k8-opt/bin/external/x11" "-iquote" "external/spdlog" "-iquote" "bazel-out/k8-opt/bin/external/spdlog" "-iquote" "external/tinyxml2" "-iquote" "bazel-out/k8-opt/bin/external/tinyxml2" "-iquote" "external/yaml_cpp" "-iquote" "bazel-out/k8-opt/bin/external/yaml_cpp" "-D" "__MKDOC_PY__" "-D" "DRAKE_COMMON_SYMBOLIC_DETAIL_HEADER" "-D" "EIGEN_MPL2_ONLY" "-D" "HAVE_CSTDDEF" "-D" "FMT_HEADER_ONLY=1" "-D" "FMT_NO_FMT_STRING_ALIAS=1" "-D" "SPDLOG_COMPILED_LIB" "-D" "HAVE_SPDLOG" "-I" "bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/lcmtypes_drake_cc" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/acrobot" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/allegro" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/call_python" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/point" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/point_cloud" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/header" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/quaternion" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/image" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/image_array" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/contact_results_for_viz" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/hydroelastic_contact_surface_for_viz" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/hydroelastic_contact_surface_tri_for_viz" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/hydroelastic_quadrature_per_point_data_for_viz" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/point_pair_contact_info_for_viz" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/drake_signal" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/iiwa" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/jaco" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/panda" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/planar_manipuland_status" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/planar_gripper" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/robot_plan" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/schunk" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/scope" "-I" "bazel-out/k8-opt/bin/lcmtypes/_virtual_includes/viewer" "-I" "bazel-out/k8-opt/bin/common/_virtual_includes/drake_marker_shared_library" "-internal-isystem" "/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++" "-internal-isystem" "/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/x86_64-linux-gnu" "-internal-isystem" "/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/backward" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/usr/include/clang/9.0.1/include/" "-internal-externc-isystem" "/usr/include/x86_64-linux-gnu" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-w" "-std=c++17" "-fdeprecated-macro" "-fdebug-compilation-dir" "..." "-ferror-limit" "19" "-fmessage-length" "0" "-fobjc-runtime=gcc" "-fcxx-exceptions" "-fexceptions" "-fdiagnostics-show-option" "-fno-spell-checking" "-faddrsig" "-x" "c++" "foo"
Traceback (most recent call last):
  File "/home/jwnimmer/tmp/bungbung/9bcc9bda03176437b32e37b105fb5071/sandbox/processwrapper-sandbox/59/execroot/drake/bazel-out/host/bin/tools/workspace/pybind11/mkdoc.runfiles/drake/tools/workspace/pybind11/mkdoc.py", line 858, in <module>
    main()
  File "/home/jwnimmer/tmp/bungbung/9bcc9bda03176437b32e37b105fb5071/sandbox/processwrapper-sandbox/59/execroot/drake/bazel-out/host/bin/tools/workspace/pybind11/mkdoc.runfiles/drake/tools/workspace/pybind11/mkdoc.py", line 810, in main
    raise RuntimeError(
RuntimeError: The operating system's C++ standard library is not installed correctly.
----------------
Note: The failure of target //tools/workspace/pybind11:mkdoc (with exit code 1) may have been caused by the fact that it is running under Python 3 instead of Python 2. Examine the error to determine if that appears to be the problem. Since this target is built in the host configuration, the only way to change its version is to set --host_force_python=PY2, which affects the entire build.

If this error started occurring in Bazel 0.27 and later, it may be because the Python toolchain now enforces that targets analyzed as PY2 and PY3 run under a Python 2 and Python 3 interpreter, respectively. See https://github.com/bazelbuild/bazel/issues/7899 for more information.
----------------
Target //bindings/pydrake:documentation_pybind.h failed to build
```

--- a/tools/workspace/pybind11/mkdoc.py
+++ b/tools/workspace/pybind11/mkdoc.py
@@ -777,6 +777,11 @@ def main():
     os.mkdir(tmpdir)
     glue_filename = os.path.join(tmpdir, "mkdoc_glue.h")
     with open(glue_filename, 'w') as glue_f:
+        # As the first line of the glue file, include a C++17 standard library
+        # file to sanity check that it's working, before we start processing
+        # the user headers.
+        glue_f.write("#include <optional>\n")
+        # Add the includes to the glue, and as comments in the output.
         for include_file in sorted(include_files):
             line = "#include \"{}\"".format(include_file)
             glue_f.write(line + "\n")
@@ -793,6 +798,18 @@ def main():
         if not translation_unit:
             raise RuntimeError(
                 "Parsing headers using the clang library failed")
+        # If there is an error on line 1, that means the C++ standard library
+        # include paths are broken.
+        if translation_unit.diagnostics:
+            if translation_unit.diagnostics[0].location.line == 1:
+                try:
+                    # Use '###' to dump Clang's include paths to stdout.
+                    index.parse("foo", parameters + ["-###"])
+                except Exception:
+                    pass
+                raise RuntimeError(
+                    ("The operating system's C++ standard library is not "
+                     "installed correctly."))
         severities = [
             diagnostic.severity for diagnostic in translation_unit.diagnostics
             if diagnostic.severity >= cindex.Diagnostic.Error

